# qc_mags

[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A521.04.0-23aa62.svg?labelColor=000000)](https://www.nextflow.io/)
[![run with docker](https://img.shields.io/badge/run%20with-docker-0db7ed?labelColor=000000&logo=docker)](https://www.docker.com/)
[![run with singularity](https://img.shields.io/badge/run%20with-singularity-1d355c.svg?labelColor=000000)](https://sylabs.io/docs/)

This workflow performs assembly quality control, with a focus on metagenome-assembled genomes (MAGs).

## Pipeline Summary

It runs the following steps on the assembly (`fasta` files):

1. **gtdbtk classify_wf**: Assemblies are classified using steps detailed in the [gtdbtk docs](https://ecogenomics.github.io/GTDBTk/commands/classify_wf.html). The `ani_screen` step is skipped.
2. **checkm2 and GUNC**: [CheckM2](https://github.com/chklovski/CheckM2) is used to predict the completeness and contamination of genomic bins (independent of their taxonomic classification). [GUNC](https://github.com/grp-bork/gunc) is also run to check for chimerism and contamination.
3. **MDMCleaner**: [MDMCleaner](https://github.com/KIT-IBG-5/mdmcleaner) is used to remove contaminants and provide reliable contig classification of metagenome assembled genomes (MAGs). Sequences that are less than 1000 bases long are removed from the resulting filtered/decontaminated fastas using [seqkit](https://bioinf.shenwei.me/seqkit/).
4. **checkm2 and GUNC**: The same checkm2 and GUNC commands (stage 2) are run on the seqkit fasta outputs.
5. **Reporting** A summary CSV report is created that combines the summary files of checkm2, gtdbtk and GUNC runs (both before and after decontamination with MDMCleaner).

### Parameters

Several parameters set paths to various databases used by pipeline tools:

- `--checkm2_db` (default: `/data/pam/software/checkm2_db/uniref100.KO.1.dmnd`)
- `--gunc_db` (default: `/data/pam/software/gunc/GTDB/gunc_db_gtdb95.dmnd`)
- `--mdmcleaner_db` (default: `/data/pam/software/mdmcleaner/v1`)
- `--gtdbtk_db` (default: `/data/pam/software/GTDBTk/release226`)

Other parameters:

- `--fasta_ext` (default: `.fa`): Set the expected extension of fasta files in the input directories.
- `--report_config` (default: [`./assets/report_config.json`](./assets/report_config.json)). See [configuration](#configuration) for more details.

## Report configuration

A configuration file can be supplied to the `--report_config` option to customise the final (per-sample) report generated by the pipeline. In particular, it allows (per-tool) customization of the identity column (to join the tool report tables together) and extract particular columns from these reports. The following tool names are currently mandatory: `GTDBTK`, `GUNC`, `CHECKM2` (NOTE: tool name must be uppercase). In the following example config block, we select 2 columns (`classification` and `closest_genome_reference`) to extract from the GTDB-Tk report and to use `user_genome` as the identity column:

```
"GTDBTK": {
    "id_column": "user_genome",
    "keep_columns": [
        "classification",
        "closest_genome_reference"
    ]
}
```

Please refer to the default [`report_config.json`](./assets/report_config.json) for expected JSON structure.

NOTE: In addition to columns derived from the tool reports, the script includes 4 columns `preqc_genome_name`, `postqc_genome_name`, `sample_or_strain_name` and `genome_status`. These are currently all derived from filenames (at some point).

| column                | description                                                                              |
| --------------------- | ---------------------------------------------------------------------------------------- |
| preqc_genome_name     | Name of the input fasta file minus extension                                             |
| post_genome_name      | Name of the fasta file output after processing with `seqkit` minus extension             |
| sample_or_strain_name | Name of the fasta file up to the last `_` character                                      |
| genome_status         | `mag` if the fasta file contains the string `mag` (lower/uppercase), `isolate` otherwise |

### Inputs

- (Metagenomic) assemblies per sample. Channel emitting elements `[meta, [1.fa, 2.fa, ...]]`, where meta is a map containing metadata (including the mandatory `ID`, tracking a sample ID).

### Outputs

- (Binned) assemblies, each corresponding to a taxonomic unit, that have been decontaminated using MDMCleaner. These decontaminated assemblies are also filtered (sequences less than 1000 bases long are removed). Both sets of assemblies are retained.
- Reports generated by CheckM2, GTDB-Tk, GUNC.
- Summary report TSV per-sample (with customisable columns) combining individual reports from various pipeline tools.

### Dependencies

All dependencies are containerised. The [report.py](./bin/report.py) is used for the `REPORTING` process to generate the TSV summary.
